distributed/scheduler.py

```diff
@@ -2651,23 +2697,32 @@ def coerce_address(self, addr):
             addr = '%s:%d' % (ip, port)
         return addr
 
-    def workers_list(self, workers):
-        """
-        List of qualifying workers
+    def workers_set(self, workers):
+        """Set of qualifying workers.
 
-        Takes a list of worker addresses or hostnames.
-        Returns a list of all worker addresses that match
+        Takes a list of worker addresses, aliases, ip's, hostnames, or
+        environment names. Returns a list of all worker addresses that match
         """
         if workers is None:
-            return list(self.ncores)
+            return set(self.ncores)
 
         out = set()
         for w in workers:
-            if ':' in w:
-                out.add(w)
+            if isinstance(w, list):
+                w = tuple(w)
+            if w in self.environment_workers:
+                out.update(self.environment_workers[w])
             else:
-                out.update({ww for ww in self.ncores if w in ww}) # TODO: quadratic
-        return list(out)
+                try:
+                    w = self.coerce_address(w)
+                except Exception:
+                    continue
+                if ':' in w:
+                    out.add(w)
+                else:
+                    out.update('%s:%s' % (w, port)
+                               for port in self.host_info[w].get('ports', ()))
+        return out
```

Comment:
Yes, I can see how the hostname environment-name ambiguity will start to hurt here